pkg/limits/stream_metadata.go

```diff
@@ -103,76 +108,143 @@ func (s *streamMetadata) Usage(tenant string, fn UsageFunc) {
 }
-func (s *streamMetadata) Store(tenant string, partitionID int32, streamHash, recTotalSize uint64, recordTime, bucketStart, bucketCutOff int64) {
+
+func (s *streamMetadata) StoreIf(tenant string, streams map[int32][]Stream, maxActiveStreams uint64, cutoff, bucketStart, bucketCutOff int64) (map[Reason][]uint64, uint64) {
 	i := s.getStripeIdx(tenant)
 
 	s.locks[i].Lock()
 	defer s.locks[i].Unlock()
 
-	// Initialize tenant map if it doesn't exist
 	if _, ok := s.stripes[i][tenant]; !ok {
-		s.stripes[i][tenant] = make(map[int32][]Stream)
+		s.stripes[i][tenant] = make(map[int32]map[uint64]Stream)
 	}
 
-	// Initialize partition map if it doesn't exist
-	if s.stripes[i][tenant][partitionID] == nil {
-		s.stripes[i][tenant][partitionID] = make([]Stream, 0)
-	}
+	var (
+		exceedLimits  = make(map[Reason][]uint64)
+		ingestedBytes uint64
+	)
+	for partitionID, streams := range streams {
+		if _, ok := s.stripes[i][tenant][partitionID]; !ok {
+			s.stripes[i][tenant][partitionID] = make(map[uint64]Stream)
+		}
+
+		var (
+			activeStreams = 0
+			newStreams    = 0
+		)
+
+		// Count as active streams all stream that are not expired.
+		for _, stored := range s.stripes[i][tenant][partitionID] {
+			if stored.LastSeenAt >= cutoff {
+				activeStreams++
+			}
+		}
 
-	for j, stream := range s.stripes[i][tenant][partitionID] {
-		if stream.Hash == streamHash {
-			// Update total size
-			totalSize := stream.TotalSize + recTotalSize
+		for _, stream := range streams {
+			stored, found := s.stripes[i][tenant][partitionID][stream.Hash]
 
-			// Update or add size for the current bucket
-			updated := false
-			sb := make([]RateBucket, 0, len(stream.RateBuckets)+1)
+			// If the stream is new or expired, check if it exceeds the limit.
+			// If limit is not exceeded and the stream is expired, reset the stream.
+			if !found || (stored.LastSeenAt < cutoff) {
+				// Count up the new stream before updating
+				newStreams++
 
-			// Only keep buckets within the rate window and update the current bucket
-			for _, bucket := range stream.RateBuckets {
-				// Clean up buckets outside the rate window
-				if bucket.Timestamp < bucketCutOff {
+				// Drop streams that exceed the limit
+				if activeStreams+newStreams > int(maxActiveStreams) {
+					exceedLimits[ReasonExceedsMaxStreams] = append(exceedLimits[ReasonExceedsMaxStreams], stream.Hash)
 					continue
 				}
 
-				if bucket.Timestamp == bucketStart {
-					// Update existing bucket
-					sb = append(sb, RateBucket{
-						Timestamp: bucketStart,
-						Size:      bucket.Size + recTotalSize,
-					})
-					updated = true
-				} else {
-					// Keep other buckets within the rate window as is
-					sb = append(sb, bucket)
+				// If the stream is stored and expired, reset the stream
+				if found && stored.LastSeenAt < cutoff {
+					stored.LastSeenAt = stream.LastSeenAt
+					stored.TotalSize = 0
+					stored.RateBuckets = []RateBucket{}
+					s.stripes[i][tenant][partitionID][stream.Hash] = stored
```

Comment:
I would just store a new value here 👍 

```suggestion
					s.stripes[i][tenant][partitionID][stream.Hash] = Stream{LastSeenAt: stream.LastSeenAt}
```